src/common/ad_tx.c

```diff
@@ -146,18 +161,62 @@ ad_tx_end(struct ad_tx *tx, int err)
 	int	rc = 0;
 
 	if (err)
-		rc = ad_tx_act_replay(&tx->tx_undo);
+		rc = ad_tx_act_replay(tx, &tx->tx_undo);
 
 	/* TODO write actions in redo list to WAL */
-
+	rc = tx_complete(tx, err);
 	if (rc == 0) {
 		ad_tx_act_cleanup(&tx->tx_undo);
 		ad_tx_act_cleanup(&tx->tx_redo);
 	}
-
+	blob_decref(tx->tx_blob);
 	return rc;
 }
 
+/**
+ * snapshot data from address to either redo or undo log
+ */
+int
+ad_tx_snap(struct ad_tx *tx, void *addr, daos_size_t size, uint32_t flags)
+{
+	bool	 undo = (flags & AD_TX_UNDO);
+	bool	 redo = (flags & AD_TX_REDO);
+
+	if (redo == undo)
+		return -DER_INVAL;
+
+	if (addr == NULL || size == 0 || size > UMEM_ACT_PAYLOAD_MAX_LEN)
+		return -DER_INVAL;
+
+	if (tx == NULL) /* noop */
+		return 0;
+
+	if (undo) {
+		struct umem_act_item	*it_undo;
+
+		D_ALLOC_ACT(it_undo, UMEM_ACT_COPY, size);
+		if (it_undo == NULL)
+			return -DER_NOMEM;
+
+		act_copy_payload(&it_undo->it_act, addr, size);
+		it_undo->it_act.ac_copy.addr = blob_ptr2addr(tx->tx_blob, addr);
+		it_undo->it_act.ac_copy.size = size;
+		AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
+	} else {
+		struct umem_act_item	*it_redo = NULL;
+
+		D_ALLOC_ACT(it_redo, UMEM_ACT_COPY, size);
+		if (it_redo == NULL)
+			return -DER_NOMEM;
+
+		act_copy_payload(&it_redo->it_act, addr, size);
+		it_redo->it_act.ac_copy.addr = blob_ptr2addr(tx->tx_blob, addr);
+		it_redo->it_act.ac_copy.size = size;
+		AD_TX_ACT_ADD(tx, it_redo, ACT_REDO);
```

Comment:
[style] Above codes could be cleaned to share, only different is last line.
AD_TX_ACT_ADD(tx, act_item, undo ? ACT_UNDO : ACT_REDO)